File: backend/app/routes/exports.py
from __future__ import annotations

from datetime import date

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from ..core.auth import require_actor
from ..db import get_db
from ..services.export_service import (
    ExportError,
    export_aging,
    export_expiry_batches,
    export_inventory,
    export_inventory_analysis,
    export_product,
    export_sales,
)
from ..services.permission_service import PermissionDenied

router = APIRouter(prefix="/exports", tags=["exports"])

XLSX_MEDIA_TYPE = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"


def _handle(exc: Exception):
    if isinstance(exc, PermissionDenied):
        raise HTTPException(status_code=403, detail=str(exc))
    if isinstance(exc, ExportError):
        raise HTTPException(status_code=400, detail=str(exc))
    raise exc


def _xlsx_response(content: bytes, filename: str, row_count: int | None = None, ascii_filename: str = "export.xlsx"):
    from urllib.parse import quote

    from fastapi.responses import Response

    # RFC 6266/5987：ASCII 文件名兜底 + filename* 携带 UTF-8 中文真实名，
    # 避免中文文件名触发 latin-1 编码失败（UnicodeEncodeError → HTTP 500）。
    safe_name = ascii_filename or "export.xlsx"
    encoded_name = quote(filename, safe="")
    content_disposition = f"attachment; filename=\"{safe_name}\"; filename*=UTF-8''{encoded_name}"
    headers = {"Content-Disposition": content_disposition}
    if row_count is not None:
        headers["X-Export-Row-Count"] = str(row_count)
    return Response(
        content=content,
        media_type=XLSX_MEDIA_TYPE,
        headers=headers,
    )


@router.get("/product")
def download_product(department_code: str = "B2C", actor: dict = Depends(require_actor), db: Session = Depends(get_db)):
    try:
        result = export_product(db, actor["user_id"], department_code)
        return _xlsx_response(result["content"], result["filename"], ascii_filename=result.get("ascii_filename", "export.xlsx"))
    except Exception as exc:
        _handle(exc)


@router.get("/sales")
def download_sales(
    department_code: str = "B2C",
    start_date: date | None = None,
    end_date: date | None = None,
    days: int = 30,
    shops: str = "",
    warehouses: str = "",
    product_search: str = "",
    include_name: str = "",
    exclude_name: str = "",
    product_codes: str = "",
    actor: dict = Depends(require_actor),
    db: Session = Depends(get_db),
):
    """下载当前筛选结果对应的全部销量明细（不分页）。

    筛选条件（日期预设/店铺/仓库/商品名/商家编码等）与列表页一致，全部由后端
    基于登录用户权限范围过滤，跨部门一律 403。
    """
    try:
        result = export_sales(
            db,
            actor["user_id"],
            department_code,
            start_date=start_date,
            end_date=end_date,
            days=days,
            shops=_split_csv(shops),
            warehouses=_split_csv(warehouses),
            product_search=product_search,
            include_name=include_name,
            exclude_name=exclude_name,
            product_codes=_split_csv(product_codes),
        )
        return _xlsx_response(result["content"], result["filename"], row_count=result["row_count"], ascii_filename=result.get("ascii_filename", "export.xlsx"))
    except Exception as exc:
        _handle(exc)


def _split_csv(value: str) -> list[str]:
    if not value:
        return []
    return [x.strip() for x in value.split(",") if x.strip()]


@router.get("/inventory")
def download_inventory(department_code: str = "B2C", actor: dict = Depends(require_actor), db: Session = Depends(get_db)):
    try:
        result = export_inventory(db, actor["user_id"], department_code)
        return _xlsx_response(result["content"], result["filename"], ascii_filename=result.get("ascii_filename", "export.xlsx"))
    except Exception as exc:
        _handle(exc)


@router.get("/aging")
def download_aging(department_code: str = "B2C", actor: dict = Depends(require_actor), db: Session = Depends(get_db)):
    try:
        result = export_aging(db, actor["user_id"], department_code)
        return _xlsx_response(result["content"], result["filename"], ascii_filename=result.get("ascii_filename", "export.xlsx"))
    except Exception as exc:
        _handle(exc)


@router.get("/inventory-analysis")
def download_inventory_analysis(
    department_code: str = "B2C",
    shops: str = "",
    warehouses: str = "",
    days: int = 30,
    product_search: str = "",
    include_name: str = "",
    exclude_name: str = "",
    product_codes: str = "",
    category: str = "",
    actor: dict = Depends(require_actor),
    db: Session = Depends(get_db),
):
    """下载库存明细（库存视图的 SKU 级库存健康度），复用列表相同筛选与权限。"""
    try:
        result = export_inventory_analysis(
            db,
            actor["user_id"],
            department_code,
            shops=_split_csv(shops),
            warehouses=_split_csv(warehouses),
            days=days,
            product_search=product_search,
            include_name=include_name,
            exclude_name=exclude_name,
            product_codes=_split_csv(product_codes),
            category=category or None,
        )
        return _xlsx_response(result["content"], result["filename"], row_count=result["row_count"], ascii_filename=result.get("ascii_filename", "export.xlsx"))
    except Exception as exc:
        _handle(exc)


@router.get("/expiry-batches")
def download_expiry_batches(
    department_code: str = "B2C",
    warehouses: str = "",
    product_search: str = "",
    include_name: str = "",
    exclude_name: str = "",
    product_codes: str = "",
    statuses: str = "",
    remaining_days_min: int | None = None,
    remaining_days_max: int | None = None,
    remaining_pct_min: float | None = None,
    remaining_pct_max: float | None = None,
    merchant_code: str = "",
    actor: dict = Depends(require_actor),
    db: Session = Depends(get_db),
):
    """下载效期批次明细（效期视图），复用列表相同筛选与权限。"""
    try:
        result = export_expiry_batches(
            db,
            actor["user_id"],
            department_code,
            warehouses=_split_csv(warehouses),
            product_search=product_search,
            include_name=include_name,
            exclude_name=exclude_name,
            product_codes=_split_csv(product_codes),
            statuses=_split_csv(statuses),
            remaining_days_min=remaining_days_min,
            remaining_days_max=remaining_days_max,
            remaining_pct_min=remaining_pct_min,
            remaining_pct_max=remaining_pct_max,
            merchant_code=merchant_code or None,
        )
        return _xlsx_response(result["content"], result["filename"], row_count=result["row_count"], ascii_filename=result.get("ascii_filename", "export.xlsx"))
    except Exception as exc:
        _handle(exc)

File: frontend/src/api/exports.js
/**
 * 下载数据 XLSX 文件。通过 fetch 获取 blob 并触发浏览器下载。
 * @param {Object} opts { path, params }
 */
async function downloadFile({ path, params }) {
  const q = new URLSearchParams()
  Object.entries(params || {}).forEach(([k, v]) => {
    if (v === undefined || v === null || v === '') return
    q.set(k, String(v))
  })
  const suffix = q.toString() ? `?${q}` : ''
  const res = await fetch(`${path}${suffix}`, { method: 'GET', credentials: 'include' })
  if (res.status === 401) window.dispatchEvent(new CustomEvent('bjr:auth-required'))
  if (!res.ok) {
    let detail = `HTTP ${res.status}`
    try {
      const p = await res.json()
      detail = typeof p.detail === 'string' ? p.detail : JSON.stringify(p.detail || p)
    } catch { /* ignore */ }
    const err = new Error(detail)
    err.status = res.status
    throw err
  }
  const disposition = res.headers.get('Content-Disposition') || ''
  // 优先解析 filename*=UTF-8''...（中文真实文件名），缺失时回退 ASCII filename。
  const starMatch = disposition.match(/filename\*=UTF-8''([^;]+)/i)
  let filename = 'download.xlsx'
  if (starMatch) {
    try { filename = decodeURIComponent(starMatch[1]) } catch { filename = starMatch[1] }
  } else {
    const match = disposition.match(/filename="?([^";]+)"?/)
    filename = match ? match[1] : 'download.xlsx'
  }
  const rowCount = Number(res.headers.get('X-Export-Row-Count') || 0) || 0
  const blob = await res.blob()
  const url = URL.createObjectURL(blob)
  const a = document.createElement('a')
  a.href = url
  a.download = filename
  document.body.appendChild(a)
  a.click()
  a.remove()
  URL.revokeObjectURL(url)
  return { filename, rowCount }
}

export function exportProduct(departmentCode) {
  return downloadFile({ path: '/api/exports/product', params: { department_code: departmentCode } })
}

export function exportSales({ departmentCode, startDate, endDate, days, shops, warehouses, productSearch, includeName, excludeName, productCodes }) {
  return downloadFile({
    path: '/api/exports/sales',
    params: {
      department_code: departmentCode,
      start_date: startDate,
      end_date: endDate,
      days: days,
      shops: (shops || []).join(','),
      warehouses: (warehouses || []).join(','),
      product_search: productSearch,
      include_name: includeName,
      exclude_name: excludeName,
      product_codes: (productCodes || []).join(','),
    },
  })
}

export function exportInventory(departmentCode) {
  return downloadFile({ path: '/api/exports/inventory', params: { department_code: departmentCode } })
}

export function exportAging(departmentCode) {
  return downloadFile({ path: '/api/exports/aging', params: { department_code: departmentCode } })
}

export function exportInventoryAnalysis({ departmentCode, shops, warehouses, days, productSearch, includeName, excludeName, productCodes, productCategoryIds, category, detailWarehouses }) {
  return downloadFile({
    path: '/api/exports/inventory-analysis',
    params: {
      department_code: departmentCode,
      shops: (shops || []).join(','),
      warehouses: (warehouses || []).join(','),
      days: days,
      product_search: productSearch,
      include_name: includeName,
      exclude_name: excludeName,
      product_codes: (productCodes || []).join(','),
      product_category_ids: (productCategoryIds || []).join(','),
      category: category || '',
      detail_warehouses: detailWarehouses || undefined,
    },
  })
}

export function exportExpiryBatches({ departmentCode, warehouses, productSearch, includeName, excludeName, productCodes, productCategoryIds, statuses, remainingDaysMin, remainingDaysMax, remainingPctMin, remainingPctMax }) {
  return downloadFile({
    path: '/api/exports/expiry-batches',
    params: {
      department_code: departmentCode,
      warehouses: (warehouses || []).join(','),
      product_search: productSearch,
      include_name: includeName,
      exclude_name: excludeName,
      product_codes: (productCodes || []).join(','),
      product_category_ids: (productCategoryIds || []).join(','),
      statuses: (statuses || []).join(','),
      remaining_days_min: remainingDaysMin,
      remaining_days_max: remainingDaysMax,
      remaining_pct_min: remainingPctMin,
      remaining_pct_max: remainingPctMax,
    },
  })
}
